### PR TITLE
Sema: Implicit conversion for single-expression closures of Never type (3.0)

### PR DESCRIPTION
**Pull request by slavapestov**
- Description: When we moved from the @noreturn attribute to the `Never` return type, we lost a bit of functionality. Since Never is just an ordinary type, the type checker did not allow a single-expression closure to contain a Never-returning expression unless the closure type itself returned Never. This precludes using single expression closures for callbacks in certain situations. This patch adds a new rule, similar to the existing rule for dropping the return type of a single-expression closure.
- Scope of the issue: This has been affecting people in the wild while migrating code from Swift 2 to Swift 3.
- Risk: Low. The change has a bit of churn to repurpose the Void-conversion code and removes a flag; a careful reading should show the changes are equivalent.
- Tested: New test added.
- Radar: rdar://problem/28269358
- Reviewed by: @DougGregor

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -5065,6 +5073,38 @@ ClosureExpr *ExprRewriter::coerceClosureExprToVoid(ClosureExpr *closureExpr) {
...
+  if (member.is<Stmt *>()) {
+    auto returnStmt = cast<ReturnStmt>(member.get<Stmt *>());
+    auto singleExpr = returnStmt->getResult();
+
+    tc.checkIgnoredExpr(singleExpr);
+
+    SmallVector<ASTNode, 1> elements;
+    elements.push_back(singleExpr);
+
+    auto braceStmt = BraceStmt::create(tc.Context,
```
Comment (DougGregor) [inline lib/Sema/CSApply.cpp]: This kind of AST hackery feels gross, but... okay. I don't see a better localized fix.

**Review by DougGregor — APPROVED** [review]

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: It's just copy and pasted from the old code (and I know, that's a HORRIBLE excuse)

**Pull request merged**